### ISSUE TITLE
关于代码的请教

### ISSUE THREAD

**Issue by Aphroditefang** [founderfan/Data-Analysis-Open-Tasks#2]
请问是所有代码么？包括后面的模型融合

**Comment by founderfan** [comment]
是的，模型融合在最后3个Cell里面，就是取最优的线性组合

**Comment by Aphroditefang** [comment]
请问下面那句话后面fillna为什么用daty里的填充NA值啊？
dat = Read_concat_csv(file_dat, par_csv).fillna(Read_concat_csv(file_dayt, {"index_col": 0}))

**Comment by founderfan** [comment]
是这样，因为比赛第一轮的训练集X和Y是在同一个dataset里面的，但是第二轮的训练集Y是单独提供的，所以要把那一列Y补回到原来的dataset里面

**Comment by Aphroditefang** [comment]
谢谢您的耐心回答。能不能请问另一个问题呐？ 
xbin = enc.fit_transform(np.transpose(
            [das.astype("category").cat.set_categories(cat).
             cat.rename_categories(1+np.arange(len(cat))).astype("float").fillna(0).values]))[:,1:]
这里one-hot编码后，为什么要把第一列去掉呐？

**Comment by Aphroditefang** [comment]
还有就是，前面得到了dacatmap之后，为什么又进行这一步呐？

 dacatmap = pd.concat([dacatmap.iloc[:,15:20], ColS_summary(dacatmap, 
        pd.Series(["_".join([x.split("_")[i] for i in [0,2]]) for x in dacatmap.columns[:15]], index = dacatmap.columns[:15]))], axis = 1)

**Comment by founderfan** [comment]
因为onehot编码过后有共线性，比如单选题ABCD四列，如果答案不是B、C、D，那就是A，所以实际上只需要分析BCD三列就够了。

**Comment by founderfan** [comment]
我们可以把函数里的前4行拿出来跑
    dac = da.copy().replace([-np.inf, np.inf], np.nan).replace("不详", np.nan)
    dac.loc[:, "UserInfo_20"] = dac.loc[:, "UserInfo_20"].fillna(dac.loc[:, "UserInfo_19"])
    datime = pd.concat(map(lambda i: Time_to_num(dac.loc[:,i]), ictype["date"]), axis = 1)
    dacatmap = pd.concat(map(lambda i: Cat_map(dac.loc[:,ictype["catmap"][i]], ictype["catmapd"][i]), range(len(ictype["catmap"]))), axis = 1)
也就是UserInfo2, UserInfo4, UserInfo8提供了三个比较类似的城市信息（UserInfo20与这三个差异会明显一些），所以为避免共线性，我们不妨用中位数和方差对前三个城市变量进行汇总

**Comment by Aphroditefang** [comment]
哦哦，明白了，也就是生成n-1个虚拟变量哈。
daS.loc[:,cols == "max"] = daS.loc[:,cols == "max"] - daS.loc[:,cols == "median"].values
这里，为什么最大值要改为最大值-中位数呐？
dac = dac.loc[:, (tmp["freq 1"] + tmp["freq NA"])/dac.shape[0] < 1 - a] 
还有这一步，这个选择有理论依据吗？

**Comment by Aphroditefang** [comment]
dac = dac.drop(dac.columns[np.any(np.abs(np.tril(np.corrcoef(dac.rank(pct = True).fillna(0.5).values, rowvar = 0), -1)) > 0.99, axis = 0)], axis = 1)
这个去除其他多重共线性？是怎么去除的啊？

**Comment by founderfan** [comment]
用各变量的rank的相关性来算Spearman相关系数，当算出大的相关矩阵后（np.corrcoef），如果有一个变量和任何之前变量（下三角阵np.tril用来取出每一个变量和之前变量的相关系数）的Spearman相关系数绝对值高于0.99，就扔掉它
np.any(axis=1)表示每行（该变量）中的任何一列（在该变量之前的任何变量）满足条件，都要找出来删掉

**Comment by Aphroditefang** [comment]
那为什么用变量rank相关性来算相关系数呐？还有这两个问题：
daS.loc[:,cols == "max"] = daS.loc[:,cols == "max"] - daS.loc[:,cols == "median"].values
这里，为什么最大值要改为最大值-中位数呐？
dac = dac.loc[:, (tmp["freq 1"] + tmp["freq NA"])/dac.shape[0] < 1 - a]
还有这一步，这个选择有理论依据吗？

**Comment by founderfan** [comment]
因为Xgboost这种决策树类模型是二分法，只考虑相对的排序而不是具体的值
比如变量X1: [1, 2, 3], 变量X2: [1, 2, 10000], 变量Y: [1, 2, 3])
尽管X1和X2不完全共线，不论选X1还是X2，只要让X<2时Y=1，X=2是Y=2，X>2时Y=3都能拟合，所以只要两个变量如果排序关系一样的话，就会造成共线性
况且相关性会遇到极端值的影响，数据在排序意义下会更稳健

最大值减中位数相当于均值方差中的“方差”，目的是尽可能使我们制作的变量彼此独立，方便模型提取信息

缺失值和大量同样的值都会造成变量的信息稀疏，而稀疏的信息引入的噪声可能多过提供的信息（假如，给我们10000个人，询问他们“是否认识土豪朋友”，其中9990个回答"不认识"，10个回答"认识"，其中答”不认识“的人中有20%的人违约了，答”认识“的有1个人，也就是10%的人违约了，但是因为“认识”的样本太少，不足以判断这个违约率低是随机的还是真和土豪朋友有关系，与其他不那么稀疏的变量相比信息含量更少，所以在实际训练中我们会删掉这个变量）
